File: BankingNLP_v4/frontend/static/js/prompts.js
/**
 * JavaScript для управления промтами
 */

// Глобальные переменные
let promptsData = {};
let promptTypes = [];

// Инициализация при загрузке страницы
document.addEventListener('DOMContentLoaded', function() {
    loadPromptTypes();
    loadPrompts();
    setupEventListeners();
});

/**
 * Загрузить типы промтов
 */
async function loadPromptTypes() {
    try {
        const response = await fetch('/api/v1/prompts/types');
        if (response.ok) {
            promptTypes = await response.json();
            populateTypeFilters();
        } else {
            console.error('Ошибка загрузки типов промтов:', response.statusText);
        }
    } catch (error) {
        console.error('Ошибка загрузки типов промтов:', error);
    }
}

/**
 * Заполнить фильтры типами промтов
 */
function populateTypeFilters() {
    const typeFilter = document.getElementById('typeFilter');
    const createTypeSelect = document.getElementById('promptType');
    const editTypeSelect = document.getElementById('editPromptType');
    
    // Очищаем существующие опции
    typeFilter.innerHTML = '<option value="">Все типы</option>';
    createTypeSelect.innerHTML = '<option value="">Выберите тип</option>';
    editTypeSelect.innerHTML = '<option value="">Выберите тип</option>';
    
    // Добавляем типы промтов
    promptTypes.forEach(type => {
        const option = document.createElement('option');
        option.value = type;
        option.textContent = getTypeDisplayName(type);
        
        typeFilter.appendChild(option.cloneNode(true));
        createTypeSelect.appendChild(option.cloneNode(true));
        editTypeSelect.appendChild(option.cloneNode(true));
    });
}

/**
 * Получить отображаемое имя типа промта
 */
function getTypeDisplayName(type) {
    const typeNames = {
        'topic_extraction': 'Извлечение тематик',
        'product_analysis': 'Анализ продуктов',
        'sentiment_analysis': 'Анализ настроений',
        'intent_classification': 'Классификация намерений',
        'summarization': 'Суммаризация',
        'benchmark': 'Бенчмарк',
        'custom': 'Пользовательский'
    };
    return typeNames[type] || type;
}

/**
 * Загрузить список промтов
 */
async function loadPrompts() {
    try {
        const response = await fetch('/api/v1/prompts/');
        if (response.ok) {
            promptsData = await response.json();
            renderPrompts();
        } else {
            console.error('Ошибка загрузки промтов:', response.statusText);
            showAlert('Ошибка загрузки промтов', 'danger');
        }
    } catch (error) {
        console.error('Ошибка загрузки промтов:', error);
        showAlert('Ошибка загрузки промтов', 'danger');
    }
}

/**
 * Отобразить промты на странице
 */
function renderPrompts() {
    const container = document.getElementById('promptsList');
    container.innerHTML = '';
    
    const filteredPrompts = filterPrompts();
    
    if (filteredPrompts.length === 0) {
        container.innerHTML = `
            <div class="col-12">
                <div class="alert alert-info text-center">
                    <i class="fas fa-info-circle me-2"></i>
                    Промты не найдены
                </div>
            </div>
        `;
        return;
    }
    
    filteredPrompts.forEach(([promptId, prompt]) => {
        const card = createPromptCard(promptId, prompt);
        container.appendChild(card);
    });
}

/**
 * Фильтровать промты по заданным критериям
 */
function filterPrompts() {
    const typeFilter = document.getElementById('typeFilter').value;
    const statusFilter = document.getElementById('statusFilter').value;
    const searchFilter = document.getElementById('searchFilter').value.toLowerCase();
    
    return Object.entries(promptsData).filter(([promptId, prompt]) => {
        // Фильтр по типу
        if (typeFilter && prompt.type !== typeFilter) {
            return false;
        }
        
        // Фильтр по статусу
        if (statusFilter) {
            const isActive = prompt.is_active;
            if (statusFilter === 'active' && !isActive) return false;
            if (statusFilter === 'inactive' && isActive) return false;
        }
        
        // Фильтр по поиску
        if (searchFilter) {
            const searchText = `${prompt.name} ${prompt.description}`.toLowerCase();
            if (!searchText.includes(searchFilter)) {
                return false;
            }
        }
        
        return true;
    });
}

/**
 * Создать карточку промта
 */
function createPromptCard(promptId, prompt) {
    const col = document.createElement('div');
    col.className = 'col-md-6 col-lg-4 mb-4';
    
    const statusBadge = prompt.is_active 
        ? '<span class="badge bg-success">Активный</span>'
        : '<span class="badge bg-secondary">Неактивный</span>';
    
    const typeBadge = `<span class="badge bg-primary prompt-type-badge">${getTypeDisplayName(prompt.type)}</span>`;
    
    const parametersHtml = prompt.parameters.length > 0 
        ? prompt.parameters.map(param => `<span class="badge parameter-badge">${param}</span>`).join('')
        : '<span class="text-muted">Нет параметров</span>';
    
    col.innerHTML = `
        <div class="card prompt-card h-100">
            <div class="card-header d-flex justify-content-between align-items-start">
                <div>
                    <h6 class="card-title mb-1">${prompt.name}</h6>
                    <small class="text-muted">v${prompt.version}</small>
                </div>
                <div>
                    ${statusBadge}
                    ${typeBadge}
                </div>
            </div>
            <div class="card-body">
                <p class="card-text">${prompt.description || 'Описание отсутствует'}</p>
                
                <div class="mb-3">
                    <strong>Параметры:</strong><br>
                    ${parametersHtml}
                </div>
                
                <div class="mb-3">
                    <strong>Шаблон:</strong>
                    <div class="prompt-template">
                        ${escapeHtml(prompt.template.substring(0, 200))}${prompt.template.length > 200 ? '...' : ''}
                    </div>
                </div>
            </div>
            <div class="card-footer">
                <div class="btn-group w-100" role="group">
                    <button class="btn btn-outline-primary btn-sm" onclick="editPrompt('${promptId}')">
                        <i class="fas fa-edit me-1"></i>Редактировать
                    </button>
                    <button class="btn btn-outline-info btn-sm" onclick="formatPrompt('${promptId}')">
                        <i class="fas fa-code me-1"></i>Форматировать
                    </button>
                    <button class="btn btn-outline-${prompt.is_active ? 'warning' : 'success'} btn-sm" 
                            onclick="togglePromptStatus('${promptId}', ${prompt.is_active})">
                        <i class="fas fa-${prompt.is_active ? 'pause' : 'play'} me-1"></i>
                        ${prompt.is_active ? 'Деактивировать' : 'Активировать'}
                    </button>
                    ${prompt.type === 'custom' ? `
                        <button class="btn btn-outline-danger btn-sm" onclick="deletePrompt('${promptId}')">
                            <i class="fas fa-trash me-1"></i>Удалить
                        </button>
                    ` : ''}
                </div>
            </div>
        </div>
    `;
    
    return col;
}

/**
 * Создать новый промт
 */
async function createPrompt() {
    const formData = {
        name: document.getElementById('promptName').value,
        description: document.getElementById('promptDescription').value,
        prompt_type: document.getElementById('promptType').value,
        template: document.getElementById('promptTemplate').value,
        parameters: document.getElementById('promptParameters').value.split(',').map(p => p.trim()).filter(p => p),
        version: document.getElementById('promptVersion').value,
        metadata: {}
    };
    
    if (!formData.name || !formData.prompt_type || !formData.template) {
        showAlert('Заполните все обязательные поля', 'warning');
        return;
    }
    
    try {
        const response = await fetch('/api/v1/prompts/', {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
            },
            body: JSON.stringify(formData)
        });
        
        if (response.ok) {
            const newPrompt = await response.json();
            showAlert('Промт успешно создан', 'success');
            bootstrap.Modal.getInstance(document.getElementById('createPromptModal')).hide();
            resetCreateForm();
            loadPrompts();
        } else {
            const error = await response.json();
            showAlert(`Ошибка создания промта: ${error.detail}`, 'danger');
        }
    } catch (error) {
        console.error('Ошибка создания промта:', error);
        showAlert('Ошибка создания промта', 'danger');
    }
}

/**
 * Редактировать промт
 */
async function editPrompt(promptId) {
    try {
        const response = await fetch(`/api/v1/prompts/${promptId}`);
        if (response.ok) {
            const prompt = await response.json();
            populateEditForm(prompt);
            bootstrap.Modal.getInstance(document.getElementById('editPromptModal')).show();
        } else {
            showAlert('Ошибка загрузки промта', 'danger');
        }
    } catch (error) {
        console.error('Ошибка загрузки промта:', error);
        showAlert('Ошибка загрузки промта', 'danger');
    }
}

/**
 * Заполнить форму редактирования
 */
function populateEditForm(prompt) {
    document.getElementById('editPromptId').value = prompt.id;
    document.getElementById('editPromptName').value = prompt.name;
    document.getElementById('editPromptDescription').value = prompt.description;
    document.getElementById('editPromptType').value = prompt.prompt_type;
    document.getElementById('editPromptTemplate').value = prompt.template;
    document.getElementById('editPromptParameters').value = prompt.parameters.join(', ');
    document.getElementById('editPromptVersion').value = prompt.version;
    document.getElementById('editPromptActive').checked = prompt.is_active;
}

/**
 * Обновить промт
 */
async function updatePrompt() {
    const promptId = document.getElementById('editPromptId').value;
    const formData = {
        name: document.getElementById('editPromptName').value,
        description: document.getElementById('editPromptDescription').value,
        template: document.getElementById('editPromptTemplate').value,
        parameters: document.getElementById('editPromptParameters').value.split(',').map(p => p.trim()).filter(p => p),
        version: document.getElementById('editPromptVersion').value,
        is_active: document.getElementById('editPromptActive').checked
    };
    
    try {
        const response = await fetch(`/api/v1/prompts/${promptId}`, {
            method: 'PUT',
            headers: {
                'Content-Type': 'application/json',
            },
            body: JSON.stringify(formData)
        });
        
        if (response.ok) {
            showAlert('Промт успешно обновлен', 'success');
            bootstrap.Modal.getInstance(document.getElementById('editPromptModal')).hide();
            loadPrompts();
        } else {
            const error = await response.json();
            showAlert(`Ошибка обновления промта: ${error.detail}`, 'danger');
        }
    } catch (error) {
        console.error('Ошибка обновления промта:', error);
        showAlert('Ошибка обновления промта', 'danger');
    }
}

/**
 * Переключить статус промта
 */
async function togglePromptStatus(promptId, currentStatus) {
    const action = currentStatus ? 'deactivate' : 'activate';
    const actionText = currentStatus ? 'деактивирован' : 'активирован';
    
    try {
        const response = await fetch(`/api/v1/prompts/${promptId}/${action}`, {
            method: 'POST'
        });
        
        if (response.ok) {
            showAlert(`Промт успешно ${actionText}`, 'success');
            loadPrompts();
        } else {
            const error = await response.json();
            showAlert(`Ошибка ${action} промта: ${error.detail}`, 'danger');
        }
    } catch (error) {
        console.error(`Ошибка ${action} промта:`, error);
        showAlert(`Ошибка ${action} промта`, 'danger');
    }
}

/**
 * Удалить промт
 */
async function deletePrompt(promptId) {
    if (!confirm('Вы уверены, что хотите удалить этот промт?')) {
        return;
    }
    
    try {
        const response = await fetch(`/api/v1/prompts/${promptId}`, {
            method: 'DELETE'
        });
        
        if (response.ok) {
            showAlert('Промт успешно удален', 'success');
            loadPrompts();
        } else {
            const error = await response.json();
            showAlert(`Ошибка удаления промта: ${error.detail}`, 'danger');
        }
    } catch (error) {
        console.error('Ошибка удаления промта:', error);
        showAlert('Ошибка удаления промта', 'danger');
    }
}

/**
 * Форматировать промт
 */
async function formatPrompt(promptId) {
    const prompt = promptsData[promptId];
    if (!prompt) return;
    
    // Простой пример параметров для демонстрации
    const parameters = {};
    prompt.parameters.forEach(param => {
        parameters[param] = `[${param}]`;
    });
    
    try {
        const response = await fetch(`/api/v1/prompts/${promptId}/format`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
            },
            body: JSON.stringify(parameters)
        });
        
        if (response.ok) {
            const result = await response.json();
            showFormattedPrompt(prompt.name, result.formatted_prompt);
        } else {
            const error = await response.json();
            showAlert(`Ошибка форматирования: ${error.detail}`, 'danger');
        }
    } catch (error) {
        console.error('Ошибка форматирования промта:', error);
        showAlert('Ошибка форматирования промта', 'danger');
    }
}

/**
 * Показать отформатированный промт
 */
function showFormattedPrompt(promptName, formattedPrompt) {
    const modal = document.createElement('div');
    modal.className = 'modal fade';
    modal.innerHTML = `
        <div class="modal-dialog modal-lg">
            <div class="modal-content">
                <div class="modal-header">
                    <h5 class="modal-title">Отформатированный промт: ${promptName}</h5>
                    <button type="button" class="btn-close" data-bs-dismiss="modal"></button>
                </div>
                <div class="modal-body">
                    <pre class="bg-light p-3 rounded" style="max-height: 400px; overflow-y: auto;">${escapeHtml(formattedPrompt)}</pre>
                </div>
                <div class="modal-footer">
                    <button type="button" class="btn btn-secondary" data-bs-dismiss="modal">Закрыть</button>
                    <button type="button" class="btn btn-primary" onclick="copyToClipboard('${escapeHtml(formattedPrompt)}')">
                        <i class="fas fa-copy me-1"></i>Копировать
                    </button>
                </div>
            </div>
        </div>
    `;
    
    document.body.appendChild(modal);
    const bootstrapModal = new bootstrap.Modal(modal);
    bootstrapModal.show();
    
    modal.addEventListener('hidden.bs.modal', () => {
        document.body.removeChild(modal);
    });
}

/**
 * Копировать в буфер обмена
 */
function copyToClipboard(text) {
    navigator.clipboard.writeText(text).then(() => {
        showAlert('Текст скопирован в буфер обмена', 'success');
    }).catch(() => {
        showAlert('Ошибка копирования', 'danger');
    });
}

/**
 * Настройка обработчиков событий
 */
function setupEventListeners() {
    // Фильтры
    document.getElementById('typeFilter').addEventListener('change', renderPrompts);
    document.getElementById('statusFilter').addEventListener('change', renderPrompts);
    document.getElementById('searchFilter').addEventListener('input', renderPrompts);
    
    // Сброс формы создания
    document.getElementById('createPromptModal').addEventListener('hidden.bs.modal', resetCreateForm);
}

/**
 * Сбросить форму создания
 */
function resetCreateForm() {
    document.getElementById('createPromptForm').reset();
    document.getElementById('promptVersion').value = '1.0';
    document.getElementById('promptActive').checked = true;
}

/**
 * Показать уведомление
 */
function showAlert(message, type) {
    const alertDiv = document.createElement('div');
    alertDiv.className = `alert alert-${type} alert-dismissible fade show position-fixed`;
    alertDiv.style.cssText = 'top: 20px; right: 20px; z-index: 9999; min-width: 300px;';
    alertDiv.innerHTML = `
        ${message}
        <button type="button" class="btn-close" data-bs-dismiss="alert"></button>
    `;
    
    document.body.appendChild(alertDiv);
    
    setTimeout(() => {
        if (alertDiv.parentNode) {
            alertDiv.remove();
        }
    }, 5000);
}

/**
 * Экранировать HTML
 */
function escapeHtml(text) {
    const div = document.createElement('div');
    div.textContent = text;
    return div.innerHTML;
} 

File: BankingNLP_v4/app/api/api_v1/endpoints/prompts.py
"""
API эндпоинты для управления промтами
"""

from typing import List, Dict, Any, Optional
from fastapi import APIRouter, HTTPException, Depends
from pydantic import BaseModel

from app.core.prompts import PromptManager, PromptTemplate, PromptType, get_prompt_manager

router = APIRouter()


class PromptCreateRequest(BaseModel):
    """Запрос на создание промта"""
    name: str
    description: str
    prompt_type: str
    template: str
    parameters: List[str] = []
    version: str = "1.0"
    metadata: Dict[str, Any] = {}


class PromptUpdateRequest(BaseModel):
    """Запрос на обновление промта"""
    name: Optional[str] = None
    description: Optional[str] = None
    template: Optional[str] = None
    parameters: Optional[List[str]] = None
    version: Optional[str] = None
    is_active: Optional[bool] = None
    metadata: Optional[Dict[str, Any]] = None


class PromptResponse(BaseModel):
    """Ответ с информацией о промте"""
    id: str
    name: str
    description: str
    prompt_type: str
    template: str
    parameters: List[str]
    version: str
    is_active: bool
    metadata: Dict[str, Any]


@router.get("/", response_model=Dict[str, Dict[str, Any]])
async def list_prompts(
    prompt_manager: PromptManager = Depends(get_prompt_manager)
):
    """
    Получить список всех промтов
    """
    try:
        return prompt_manager.list_prompts()
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Ошибка получения списка промтов: {str(e)}")


@router.get("/types", response_model=List[str])
async def get_prompt_types():
    """
    Получить список доступных типов промтов
    """
    return [pt.value for pt in PromptType]


@router.get("/type/{prompt_type}", response_model=List[PromptResponse])
async def get_prompts_by_type(
    prompt_type: str,
    prompt_manager: PromptManager = Depends(get_prompt_manager)
):
    """
    Получить промты определенного типа
    """
    try:
        pt_enum = PromptType(prompt_type)
        prompts = prompt_manager.get_prompts_by_type(pt_enum)
        
        return [
            PromptResponse(
                id=prompt_id,
                name=prompt.name,
                description=prompt.description,
                prompt_type=prompt.prompt_type.value,
                template=prompt.template,
                parameters=prompt.parameters,
                version=prompt.version,
                is_active=prompt.is_active,
                metadata=prompt.metadata
            )
            for prompt_id, prompt in prompt_manager.prompts.items()
            if prompt.prompt_type == pt_enum and prompt.is_active
        ]
    except ValueError:
        raise HTTPException(status_code=400, detail=f"Неизвестный тип промта: {prompt_type}")
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Ошибка получения промтов: {str(e)}")


@router.get("/{prompt_id}", response_model=PromptResponse)
async def get_prompt(
    prompt_id: str,
    prompt_manager: PromptManager = Depends(get_prompt_manager)
):
    """
    Получить промт по ID
    """
    try:
        prompt = prompt_manager.get_prompt(prompt_id)
        if not prompt:
            raise HTTPException(status_code=404, detail=f"Промт {prompt_id} не найден")
        
        return PromptResponse(
            id=prompt_id,
            name=prompt.name,
            description=prompt.description,
            prompt_type=prompt.prompt_type.value,
            template=prompt.template,
            parameters=prompt.parameters,
            version=prompt.version,
            is_active=prompt.is_active,
            metadata=prompt.metadata
        )
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Ошибка получения промта: {str(e)}")


@router.post("/", response_model=PromptResponse)
async def create_prompt(
    request: PromptCreateRequest,
    prompt_manager: PromptManager = Depends(get_prompt_manager)
):
    """
    Создать новый промт
    """
    try:
        # Генерируем ID для нового промта
        prompt_id = f"custom_{request.name.lower().replace(' ', '_')}_{request.version}"
        
        # Проверяем, что промт с таким ID не существует
        if prompt_manager.get_prompt(prompt_id):
            raise HTTPException(status_code=400, detail=f"Промт с ID {prompt_id} уже существует")
        
        # Создаем промт
        prompt = PromptTemplate(
            name=request.name,
            description=request.description,
            prompt_type=PromptType(request.prompt_type),
            template=request.template,
            parameters=request.parameters,
            version=request.version,
            metadata=request.metadata
        )
        
        prompt_manager.add_prompt(prompt_id, prompt)
        
        return PromptResponse(
            id=prompt_id,
            name=prompt.name,
            description=prompt.description,
            prompt_type=prompt.prompt_type.value,
            template=prompt.template,
            parameters=prompt.parameters,
            version=prompt.version,
            is_active=prompt.is_active,
            metadata=prompt.metadata
        )
    except ValueError as e:
        raise HTTPException(status_code=400, detail=f"Ошибка валидации: {str(e)}")
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Ошибка создания промта: {str(e)}")


@router.put("/{prompt_id}", response_model=PromptResponse)
async def update_prompt(
    prompt_id: str,
    request: PromptUpdateRequest,
    prompt_manager: PromptManager = Depends(get_prompt_manager)
):
    """
    Обновить существующий промт
    """
    try:
        # Проверяем, что промт существует
        prompt = prompt_manager.get_prompt(prompt_id)
        if not prompt:
            raise HTTPException(status_code=404, detail=f"Промт {prompt_id} не найден")
        
        # Подготавливаем данные для обновления
        update_data = {}
        if request.name is not None:
            update_data['name'] = request.name
        if request.description is not None:
            update_data['description'] = request.description
        if request.template is not None:
            update_data['template'] = request.template
        if request.parameters is not None:
            update_data['parameters'] = request.parameters
        if request.version is not None:
            update_data['version'] = request.version
        if request.is_active is not None:
            update_data['is_active'] = request.is_active
        if request.metadata is not None:
            update_data['metadata'] = request.metadata
        
        prompt_manager.update_prompt(prompt_id, **update_data)
        
        # Получаем обновленный промт
        updated_prompt = prompt_manager.get_prompt(prompt_id)
        
        return PromptResponse(
            id=prompt_id,
            name=updated_prompt.name,
            description=updated_prompt.description,
            prompt_type=updated_prompt.prompt_type.value,
            template=updated_prompt.template,
            parameters=updated_prompt.parameters,
            version=updated_prompt.version,
            is_active=updated_prompt.is_active,
            metadata=updated_prompt.metadata
        )
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Ошибка обновления промта: {str(e)}")


@router.delete("/{prompt_id}")
async def delete_prompt(
    prompt_id: str,
    prompt_manager: PromptManager = Depends(get_prompt_manager)
):
    """
    Удалить промт
    """
    try:
        prompt_manager.delete_prompt(prompt_id)
        return {"message": f"Промт {prompt_id} успешно удален"}
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Ошибка удаления промта: {str(e)}")


@router.post("/{prompt_id}/format")
async def format_prompt(
    prompt_id: str,
    parameters: Dict[str, Any],
    prompt_manager: PromptManager = Depends(get_prompt_manager)
):
    """
    Форматировать промт с переданными параметрами
    """
    try:
        formatted_prompt = prompt_manager.format_prompt(prompt_id, **parameters)
        return {"formatted_prompt": formatted_prompt}
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Ошибка форматирования промта: {str(e)}")


@router.post("/{prompt_id}/activate")
async def activate_prompt(
    prompt_id: str,
    prompt_manager: PromptManager = Depends(get_prompt_manager)
):
    """
    Активировать промт
    """
    try:
        prompt_manager.update_prompt(prompt_id, is_active=True)
        return {"message": f"Промт {prompt_id} активирован"}
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Ошибка активации промта: {str(e)}")


@router.post("/{prompt_id}/deactivate")
async def deactivate_prompt(
    prompt_id: str,
    prompt_manager: PromptManager = Depends(get_prompt_manager)
):
    """
    Деактивировать промт
    """
    try:
        prompt_manager.update_prompt(prompt_id, is_active=False)
        return {"message": f"Промт {prompt_id} деактивирован"}
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Ошибка деактивации промта: {str(e)}") 